src/gdal/ogr_file_format.cpp

```diff
@@ -1405,5 +1379,573 @@ LatLon OgrFileImport::calcAverageLatLon(OGRDataSourceH data_source)
+		sym_name.truncate(32);
+
+		auto line_string = ogr::unique_geometry(OGR_G_CreateGeometry(wkbLineString));
+		const auto& parts = path->parts();
+		for (const auto& part : parts)
+		{
+			auto po_feature = ogr::unique_feature(OGR_F_Create(OGR_L_GetLayerDefn(layer)));
+			OGR_F_SetFieldString(po_feature.get(), OGR_F_GetFieldIndex(po_feature.get(), "Name"), sym_name.toLatin1().constData());
+
+			for (const auto& coord : part.path_coords)
+			{
+				QPointF proj_cord = georef.toProjectedCoords(coord.pos);
+				OGR_G_AddPoint_2D(line_string.get(), proj_cord.x(), proj_cord.y());
+			}
+
+			if (need_wgs1984)
+				OGR_G_Transform(line_string.get(), transformation.get());
+
+			OGR_F_SetGeometry(po_feature.get(), line_string.get());
+
+			OGR_F_SetStyleString(po_feature.get(),
+			                     OGR_STBL_Find(table.get(), QString::number(map->findSymbolIndex(symbol)).toLatin1().constData()));
+
+			if (OGR_L_CreateFeature(layer, po_feature.get()) != OGRERR_NONE)
+				throw FileFormatException(tr("Failed to create line feature in layer"));
+		}
+	};
+
+	map->applyOnMatchingObjects(add_feature, condition);
+}
+
+void OgrFileExport::addAreasToLayer(OGRLayerH layer, const std::function<bool (const Object*)>& condition)
+{
+	const auto& georef = map->getGeoreferencing();
+
+	auto add_feature = [&](const Object* object) {
+		const auto& symbol = object->getSymbol();
+		const auto& path = object->asPath();
+		if (path->parts().size() < 1)
+			return;
+
+		auto po_feature = ogr::unique_feature(OGR_F_Create(OGR_L_GetLayerDefn(layer)));
+
+		QString sym_name = symbol->getPlainTextName();
+		sym_name.truncate(32);
+		OGR_F_SetFieldString(po_feature.get(), OGR_F_GetFieldIndex(po_feature.get(), "Name"), sym_name.toLatin1().constData());
+
+		auto polygon = ogr::unique_geometry(OGR_G_CreateGeometry(wkbPolygon));
+		auto cur_ring = ogr::unique_geometry(OGR_G_CreateGeometry(wkbLinearRing));
+
+		const auto& parts = path->parts();
+		for (const auto& part : parts)
+		{
+			for (const auto& coord : part.path_coords)
+			{
+				QPointF proj_cord = georef.toProjectedCoords(coord.pos);
+				OGR_G_AddPoint_2D(cur_ring.get(), proj_cord.x(), proj_cord.y());
+			}
+			OGR_G_CloseRings(cur_ring.get());
+			if (need_wgs1984)
+				OGR_G_Transform(cur_ring.get(), transformation.get());
+			OGR_G_AddGeometry(polygon.get(), cur_ring.get());
+			cur_ring.reset(OGR_G_CreateGeometry(wkbLinearRing));
+		}
+
+		OGR_F_SetGeometry(po_feature.get(), polygon.get());
+
+		OGR_F_SetStyleString(po_feature.get(),
+		                     OGR_STBL_Find(table.get(), QString::number(map->findSymbolIndex(symbol)).toLatin1().constData()));
+
+		if (OGR_L_CreateFeature(layer, po_feature.get()) != OGRERR_NONE)
+			throw FileFormatException(tr("Failed to create feature in layer"));
+	};
+
+	map->applyOnMatchingObjects(add_feature, condition);
+}
+
+OGRLayerH OgrFileExport::createLayer(const char* layer_name, OGRwkbGeometryType type)
+{
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,0,0)
```

Comment:
This is great. It should replace the `#ifdef GDAL_DMD_EXTENSIONS` in the existing code.